### PR TITLE
Informed choice of yaml provider for root transforms.

### PR DESCRIPTION
**Pull request by robertwb**
Beam now considers the possible providers of consuming transforms. There is still room for improvement, but this should be a decent general heuristic.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by robertwb** [comment]
R: @bzablocki could you review this as you already know a bit about the provider affinity concept?

File: sdks/python/apache_beam/yaml/yaml_transform.py [hunk 1]
```diff
@@ -208,6 +220,56 @@ def get_outputs(self, transform_name):
...
+    if input_providers:
+      possible_providers = best_matches(
+          possible_providers, [[p] for p in input_providers])
+
+    # Try to match downstream operations, continuing until there is no tie.
+    adjacent_transforms = [spec['__uuid__']]
+    while len(possible_providers) > 1:
+      # Go downstream one step. (This is why we start with spec itself.)
+      adjacent_transforms = sum(
+          [list(self.followers(t)) for t in adjacent_transforms], [])
```
Comment (bzablocki) [inline sdks/python/apache_beam/yaml/yaml_transform.py:261]: `adjacent_transforms` is always a list of length 1, why do we have to keep it as a list and iterate through it?

File: sdks/python/apache_beam/yaml/yaml_transform.py [hunk 2]
```diff
@@ -208,6 +220,56 @@ def get_outputs(self, transform_name):
...
+          'No available provider for type %r at %s' %
+          (spec['type'], identify_object(spec)))
+
+    # Only one choice, no need to rank further.
+    if len(possible_providers) == 1:
+      return possible_providers[0]
+
+    def best_matches(
+        possible_providers: Iterable[yaml_provider.Provider],
+        adjacent_provider_options: Iterable[Iterable[yaml_provider.Provider]]
```
Comment (bzablocki) [inline sdks/python/apache_beam/yaml/yaml_transform.py:242]: Can we lower the number of dimensions somehow and not have nested lists?

File: sdks/python/apache_beam/yaml/yaml_transform.py [hunk 3]
```diff
@@ -208,6 +220,56 @@ def get_outputs(self, transform_name):
   def compute_outputs(self, transform_id):
     return expand_transform(self._transforms_by_uuid[transform_id], self)
 
+  def best_provider(self, t, input_providers):
```
Comment (bzablocki) [inline sdks/python/apache_beam/yaml/yaml_transform.py:223]: This function is pretty complicated and uses a lot of nested lists/dicts. It's hard to keep track of what is happening without types and defined structures. Would you agree with this? Can we perhaps plan some work to refactor it a little bit?

**Review by bzablocki — COMMENTED** [review]
I must say that the implementation of this concept is pretty complex and hard to understand. I did my best to review it.

**Review by robertwb — COMMENTED** [review]
Thank you for taking a look at this.

File: sdks/python/apache_beam/yaml/yaml_transform.py [hunk 4]
```diff
@@ -225,19 +287,7 @@ def create_ptransform(self, spec, input_pcolls):
         providers_by_input[pcoll] for pcoll in input_pcolls
         if pcoll in providers_by_input
     ]
```
[the remark this answers is not in the document]
Comment (bzablocki) [reply sdks/python/apache_beam/yaml/yaml_transform.py:319]: ok

File: sdks/python/apache_beam/yaml/yaml_transform.py [hunk 5]
```diff
@@ -208,6 +220,56 @@ def get_outputs(self, transform_name):
...
+    def best_matches(
+        possible_providers: Iterable[yaml_provider.Provider],
+        adjacent_provider_options: Iterable[Iterable[yaml_provider.Provider]]
+    ) -> Iterable[yaml_provider.Provider]:
+      providers_by_score = collections.defaultdict(list)
+      for p in possible_providers:
+        providers_by_score[sum(
+            max(p.affinity(ap) for ap in apo)
+            for apo in adjacent_provider_options)].append(p)
+      return providers_by_score[max(providers_by_score.keys())]
```
[the remark this answers is not in the document]
Comment (bzablocki) [reply sdks/python/apache_beam/yaml/yaml_transform.py:266]: Got it.

File: sdks/python/apache_beam/yaml/yaml_transform.py (hunk 1 above)
Comment (bzablocki) [reply sdks/python/apache_beam/yaml/yaml_transform.py]: Got it,  it is more readable now, thanks.

File: sdks/python/apache_beam/yaml/yaml_transform.py (hunk 3 above)
Comment (bzablocki) [reply sdks/python/apache_beam/yaml/yaml_transform.py]: It helps, thanks

File: sdks/python/apache_beam/yaml/yaml_transform.py [hunk 6]
```diff
@@ -208,6 +229,77 @@ def get_outputs(self, transform_name):
...
+            if p.available()
+        ] for t in adjacent_transforms]
+        if any(not apo for apo in adjacent_provider_options):
+          # One of the transforms had no available providers.
+          # We will throw an error later, doesn't matter what we return.
+          break
+        # Filter down the set of possible providers to the best ones.
+        possible_providers = best_matches(
+            possible_providers, adjacent_provider_options)
+        # If we are down to one option, no ned to go further.
```
Comment (bzablocki) [inline sdks/python/apache_beam/yaml/yaml_transform.py:294]:
```suggestion
        # If we are down to one option, no need to go further.
```

**Review by bzablocki — APPROVED** [review]
Looks good, tests seem to cover all cases as well. Thanks for improving the readability.

**Comment by robertwb** [comment]
Thank you for plowing through this and your suggestions to make things more readable; definitely a better result.

**Pull request merged**